Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -281,3 +282,30 @@ lemma normEDS_even (m : ℕ) : normEDS b c d (2 * (m + 3)) * b =
 @[simp]
 lemma normEDS_neg (n : ℕ) : normEDS b c d (-n) = -normEDS b c d n := by
   rw [normEDS, preNormEDS_neg, Int.natAbs_neg, neg_mul, normEDS]
+
+/-- Strong recursion principle for a normalised EDS indexed by `ℕ`: if we have
+ * `P 0`, `P 1`, `P 2`, `P 3`, and `P 4`,
+ * for all `m : ℕ` we can prove `P (2 * (m + 3))` from `P k` for all `k < 2 * (m + 3)`, and
+ * for all `m : ℕ` we can prove `P (2 * (m + 2) + 1)` from `P k` for all `k < 2 * (m + 2) + 1`,
+then we have `P n` for all `n : ℕ`. -/
+@[elab_as_elim]
+noncomputable def normEDSRec' {P : ℕ → Sort u} (base0 : P 0) (base1 : P 1) (base2 : P 2)
+    (base3 : P 3) (base4 : P 4) (even : ∀ m : ℕ, (∀ k < 2 * (m + 3), P k) → P (2 * (m + 3)))
+    (odd : ∀ m : ℕ, (∀ k < 2 * (m + 2) + 1, P k) → P (2 * (m + 2) + 1)) (n : ℕ) :
+    P n :=
+  n.evenOddStrongRec (by rintro (_ | _ | _ | _) h; exacts [base0, base2, base4, even _ h])
+    (by rintro (_ | _ | _) h; exacts [base1, base3, odd _ h])
+
+/-- Strong recursion principle for a normalised EDS indexed by `ℤ`: if we have
+ * `P 0`, `P 1`, `P 2`, `P 3`, and `P 4`,
+ * for all `m : ℕ` we can prove `P (2 * (m + 3))` from `P k` for all `k < 2 * (m + 3)`,
+ * for all `m : ℕ` we can prove `P (2 * (m + 2) + 1)` from `P k` for all `k < 2 * (m + 2) + 1`, and
+ * for all `n : ℕ` we can extend from `P n` to `P (-n)`,
+then we have `P n` for all `n : ℤ`. -/
+@[elab_as_elim]
+noncomputable def normEDSRec {P : ℤ → Sort u} (base0 : P 0) (base1 : P 1) (base2 : P 2)
+    (base3 : P 3) (base4 : P 4) (even : ∀ m : ℕ, (∀ k < 2 * (m + 3), P k) → P (2 * (m + 3)))
+    (odd : ∀ m : ℕ, (∀ k < 2 * (m + 2) + 1, P k) → P (2 * (m + 2) + 1))
+    (neg : ∀ n : ℕ, P n → P (-n)) : ∀ n : ℤ, P n
```

Comment:
I think we are missing the recursion principle assuming `(nonneg : ∀ n : ℕ, P n) (neg : ∀ n : ℕ, P n → P (-n))`, and `normEDS` is simply the combination of this principle with `normEDS'`. This is usually achieved with `wlog : 0 ≤ n` in my PR, but it requires deducing `0 ≤ -n` from `¬ n ≤ 0`, which is annoying whether you use `linarith` or explicit lemmas like I do.

If we add this then `normEDSRec` can be removed, and the name can be used by the version I suggested, or maybe a more general version where the inductive hypotheses are like `even : ∀ m : ℕ, (∀ k : ℤ, 0 ≤ k → k < 2 * (m + 3), P k) → P (2 * (m + 3))`, which would avoid rewrites before applying the inductive hypotheses.

Reply:
Well we do have `Int.rec` that recurses on `ofNat` and `negSucc`... do you think we still want this and call it something like `Int.rec'`?

Reply:
I'm still not super convinced that we need both `normEDSRec` and `normEDSRec'`, but I've repurposed it as you said and added `Int.negInduction`. I'm somewhat disgusted by the naming inconsistency with the other recursion/induction defs in that file, so someone has to fix those in a separate PR eventually.